Adds max_id decrement to close_device

cjoster opened the pull request:
I just received an OnlyKey and while experimenting, I found an interesting bug. I haven't configured the onlykey, but I did have it inserted, while I also had my Yubikey (which is configured for u2f) inserted, and I could not login. As it turns out, every time the devices are discovered, the max_index gets incremented, only for the onlykey to fail discover, but max_id never gets decremented. So every time you devs_discover, the max_index inreases by one. Turns out it throws pam-u2f into an infinite loop, but this should be fixed here.

There's also a nasty delay while attempting to discover an unconfigured onlykey, but I'll work on that later.

Commit:
Adds max_id decrement to close_device

klali commented:
I'm not sure this is the correct fix, the variable is a bit miss-named, it should be used more as a way to tell if the device list has changed.

Consider the case where a device is inserted, so we have max_id = 1, that device is removed and another is inserted, in that case we'd still end up with 1 here.

I've pushed an alternative fix to https://github.com/Yubico/libu2f-host/tree/fixup_max_index which only increases max_id after init was succesful, but also increases it when removing a previously discovered device. Could you test that branch with your use-case?

cjoster commented:
@klali I had wondered if this was intentional or not. I'll take a look at your fix in my use case as soon as I'm back on the ground and report back.

As a side note, I'm wondering if it's appropriate to expose, "this is how far you are from the last time you asked" and instead, it should be, say, a random number or a UUID or such. I'm not sure what the vulnerability, but the question of `if( my_max_id >= devs->max_id - N)` can be asked (has there only been N changes since the last time I checked) of an incremented variable despite it not making sense. A random number would prevent an application from asking such a question.

cjoster commented:
Your proposed fix does not work...

`This tty (pts/0) is not a virtual console.


The pts/0 is now locked by lordvadr.
Password: 
debug(pam_u2f): pam-u2f.c:94 (parse_cfg): called.
debug(pam_u2f): pam-u2f.c:95 (parse_cfg): flags 0 argc 3
debug(pam_u2f): pam-u2f.c:97 (parse_cfg): argv[0]=cue
debug(pam_u2f): pam-u2f.c:97 (parse_cfg): argv[1]=authpending_file=
debug(pam_u2f): pam-u2f.c:97 (parse_cfg): argv[2]=debug
debug(pam_u2f): pam-u2f.c:99 (parse_cfg): max_devices=0
debug(pam_u2f): pam-u2f.c:100 (parse_cfg): debug=1
debug(pam_u2f): pam-u2f.c:101 (parse_cfg): interactive=0
debug(pam_u2f): pam-u2f.c:102 (parse_cfg): cue=1
debug(pam_u2f): pam-u2f.c:103 (parse_cfg): nodetect=0
debug(pam_u2f): pam-u2f.c:104 (parse_cfg): manual=0
debug(pam_u2f): pam-u2f.c:105 (parse_cfg): nouserok=0
debug(pam_u2f): pam-u2f.c:106 (parse_cfg): openasuser=0
debug(pam_u2f): pam-u2f.c:107 (parse_cfg): alwaysok=0
debug(pam_u2f): pam-u2f.c:108 (parse_cfg): authfile=(null)
debug(pam_u2f): pam-u2f.c:109 (parse_cfg): authpending_file=
debug(pam_u2f): pam-u2f.c:110 (parse_cfg): origin=(null)
debug(pam_u2f): pam-u2f.c:111 (parse_cfg): appid=(null)
debug(pam_u2f): pam-u2f.c:112 (parse_cfg): prompt=(null)
debug(pam_u2f): pam-u2f.c:157 (pam_sm_authenticate): Origin not specified, using "pam://grissom"
debug(pam_u2f): pam-u2f.c:169 (pam_sm_authenticate): Appid not specified, using the same value of origin (pam://grissom)
debug(pam_u2f): pam-u2f.c:180 (pam_sm_authenticate): Maximum devices number not set. Using default (24)
debug(pam_u2f): pam-u2f.c:198 (pam_sm_authenticate): Requesting authentication for user lordvadr
debug(pam_u2f): pam-u2f.c:209 (pam_sm_authenticate): Found user lordvadr
debug(pam_u2f): pam-u2f.c:210 (pam_sm_authenticate): Home directory for lordvadr is /home/lordvadr
debug(pam_u2f): pam-u2f.c:217 (pam_sm_authenticate): Variable XDG_CONFIG_HOME is not set. Using default value ($HOME/.config/)
debug(pam_u2f): pam-u2f.c:245 (pam_sm_authenticate): Using default authentication file /home/lordvadr/.config/Yubico/u2f_keys
debug(pam_u2f): util.c:102 (get_devices_from_authfile): Authorization line: lordvadr:KaWsrbTjPH1OB_MqvJCOw7SdeCQwbQUgx23TJ9UxYROIxSxXKyntggyccbO0sio8UYp6Ar4YrnUo6A40orkW6g,04da9404e660e720f5cca1645a54847d5ed844e1ea2366af025dcb98cd22e9fbce42104003714c3bc3f53ad42ef9993b23523bc26b564e139aa93f7769a5c26df0:uON-6lDuYfmJOeYDqI8q4ZR-eEFsRAStxKorGd6BaUV6MxIBRDB8JdGKP0j5ysNavEfVMPCPb0MwSNJ5WUB6Lw,04b1393e667f66da06041ad4081cbc0b681dee1d5e5b70dcbe95e80115d71a66423fb34ea701bf04c738a358f596a4a1ce49cd15c5e29aba38559a47daab48512f:PqSNe6j2l_ConOy4H0YmPLdVQ2zOOr6Xw5bkFp17Xh5CXOZeTv614p2FahQOYXLk2GRdWwF1L1nbxbgbB55A3g,04a0f65042cd6c3bd6e25ae7d0383cb3fdd722c3e576df892ae5706e0b2a2a7532cae55c4194db6aef489ebe44d2f6e20c8cdaca63bb74340d0af33557d1bdf447
debug(pam_u2f): util.c:107 (get_devices_from_authfile): Matched user: lordvadr
debug(pam_u2f): util.c:134 (get_devices_from_authfile): KeyHandle for device number 1: KaWsrbTjPH1OB_MqvJCOw7SdeCQwbQUgx23TJ9UxYROIxSxXKyntggyccbO0sio8UYp6Ar4YrnUo6A40orkW6g
debug(pam_u2f): util.c:153 (get_devices_from_authfile): publicKey for device number 1: 04da9404e660e720f5cca1645a54847d5ed844e1ea2366af025dcb98cd22e9fbce42104003714c3bc3f53ad42ef9993b23523bc26b564e139aa93f7769a5c26df0
debug(pam_u2f): util.c:164 (get_devices_from_authfile): Length of key number 1 is 65
debug(pam_u2f): util.c:134 (get_devices_from_authfile): KeyHandle for device number 2: uON-6lDuYfmJOeYDqI8q4ZR-eEFsRAStxKorGd6BaUV6MxIBRDB8JdGKP0j5ysNavEfVMPCPb0MwSNJ5WUB6Lw
debug(pam_u2f): util.c:153 (get_devices_from_authfile): publicKey for device number 2: 04b1393e667f66da06041ad4081cbc0b681dee1d5e5b70dcbe95e80115d71a66423fb34ea701bf04c738a358f596a4a1ce49cd15c5e29aba38559a47daab48512f
debug(pam_u2f): util.c:164 (get_devices_from_authfile): Length of key number 2 is 65
debug(pam_u2f): util.c:134 (get_devices_from_authfile): KeyHandle for device number 3: PqSNe6j2l_ConOy4H0YmPLdVQ2zOOr6Xw5bkFp17Xh5CXOZeTv614p2FahQOYXLk2GRdWwF1L1nbxbgbB55A3g
debug(pam_u2f): util.c:153 (get_devices_from_authfile): publicKey for device number 3: 04a0f65042cd6c3bd6e25ae7d0383cb3fdd722c3e576df892ae5706e0b2a2a7532cae55c4194db6aef489ebe44d2f6e20c8cdaca63bb74340d0af33557d1bdf447
debug(pam_u2f): util.c:164 (get_devices_from_authfile): Length of key number 3 is 65
debug(pam_u2f): util.c:191 (get_devices_from_authfile): Found 3 device(s) for user lordvadr
debug(pam_u2f): pam-u2f.c:310 (pam_sm_authenticate): authpending_file is set to an empty value, touch request notifications will be disabled
USB send: 00ffffffff8600080807060504030201000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000
USB write returned 65
now trying with timeout 2
now trying with timeout 4
now trying with timeout 8
now trying with timeout 16
now trying with timeout 32
now trying with timeout 64
now trying with timeout 128
now trying with timeout 256
now trying with timeout 512
now trying with timeout 1024
now trying with timeout 2048
now trying with timeout 4096
USB read rc read 64
USB send: 00ffffffff8600080807060504030201000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000
USB write returned 65
now trying with timeout 2
USB read rc read 64
USB recv: ffffffff860011080706050403020100c90005020403070100000000000000000000000000000000000000000000000000000000000000000000000000000000
device /dev/hidraw1 discovered as 'Yubikey 4 OTP+U2F+CCID'
  version (Interface, Major, Minor, Build): 2, 4, 3, 0  capFlags: 1
debug(pam_u2f): util.c:272 (do_authentication): Device max index is 1
debug(pam_u2f): util.c:306 (do_authentication): Attempting authentication with device number 1
debug(pam_u2f): util.c:330 (do_authentication): Challenge: { "keyHandle": "KaWsrbTjPH1OB_MqvJCOw7SdeCQwbQUgx23TJ9UxYROIxSxXKyntggyccbO0sio8UYp6Ar4YrnUo6A40orkW6g", "version": "U2F_V2", "challenge": "aH2Hd1Tt3nYTzphPC-8oq1ny1F28H-bwLKwPyjbiYqc", "appId": "pam:\/\/grissom" }
JSON: { "keyHandle": "KaWsrbTjPH1OB_MqvJCOw7SdeCQwbQUgx23TJ9UxYROIxSxXKyntggyccbO0sio8UYp6Ar4YrnUo6A40orkW6g", "version": "U2F_V2", "challenge": "aH2Hd1Tt3nYTzphPC-8oq1ny1F28H-bwLKwPyjbiYqc", "appId": "pam:\/\/grissom" }
JSON challenge URL-B64: aH2Hd1Tt3nYTzphPC-8oq1ny1F28H-bwLKwPyjbiYqc
client data: { "challenge": "aH2Hd1Tt3nYTzphPC-8oq1ny1F28H-bwLKwPyjbiYqc", "origin": "pam:\/\/grissom", "typ": "navigator.id.getAssertion" }
JSON: { "keyHandle": "KaWsrbTjPH1OB_MqvJCOw7SdeCQwbQUgx23TJ9UxYROIxSxXKyntggyccbO0sio8UYp6Ar4YrnUo6A40orkW6g", "version": "U2F_V2", "challenge": "aH2Hd1Tt3nYTzphPC-8oq1ny1F28H-bwLKwPyjbiYqc", "appId": "pam:\/\/grissom" }
JSON app_id pam://grissom
JSON: { "keyHandle": "KaWsrbTjPH1OB_MqvJCOw7SdeCQwbQUgx23TJ9UxYROIxSxXKyntggyccbO0sio8UYp6Ar4YrnUo6A40orkW6g", "version": "U2F_V2", "challenge": "aH2Hd1Tt3nYTzphPC-8oq1ny1F28H-bwLKwPyjbiYqc", "appId": "pam:\/\/grissom" }
JSON keyHandle URL-B64: KaWsrbTjPH1OB_MqvJCOw7SdeCQwbQUgx23TJ9UxYROIxSxXKyntggyccbO0sio8UYp6Ar4YrnUo6A40orkW6g
USB send: 0000c9000583008a00020700000081ea6edd8512262118b88616d35d5f8c6c0ff0b21943a9893bc0eb6b310faa89156240e70f034a89f360dac1a2c46c89b0a2
USB write returned 65
USB send: 0000c900050092aaf8eaac9d3076477ac75288a84029a5acadb4e33c7d4e07f32abc908ec3b49d7824306d0520c76dd327d531611388c52c572b29ed820c9c71
USB write returned 65
USB send: 0000c9000501b4b22a3c518a7a02be18ae7528e80e34a2b916ea0000000000000000000000000000000000000000000000000000000000000000000000000000
USB write returned 65
now trying with timeout 2
now trying with timeout 4
now trying with timeout 8
now trying with timeout 16
USB read rc read 64
USB recv: 00c900058300026a8000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000
USB data (len 2): 6a80
debug(pam_u2f): util.c:363 (do_authentication): Device for this keyhandle is not present.
USB send: 00ffffffff8600080807060504030201000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000
USB write returned 65
now trying with timeout 2
now trying with timeout 4
now trying with timeout 8
now trying with timeout 16
now trying with timeout 32
now trying with timeout 64
now trying with timeout 128
now trying with timeout 256
now trying with timeout 512
now trying with timeout 1024
now trying with timeout 2048
now trying with timeout 4096
USB read rc read 64
USB send: 0000c900058100010000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000
USB write returned 65
now trying with timeout 2
USB read rc read 64
USB recv: 00c90005810001000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000
debug(pam_u2f): util.c:379 (do_authentication): Devices max_index has changed: 2 (was 1). Starting over
debug(pam_u2f): util.c:306 (do_authentication): Attempting authentication with device number 1
debug(pam_u2f): util.c:330 (do_authentication): Challenge: { "keyHandle": "KaWsrbTjPH1OB_MqvJCOw7SdeCQwbQUgx23TJ9UxYROIxSxXKyntggyccbO0sio8UYp6Ar4YrnUo6A40orkW6g", "version": "U2F_V2", "challenge": "aH2Hd1Tt3nYTzphPC-8oq1ny1F28H-bwLKwPyjbiYqc", "appId": "pam:\/\/grissom" }
JSON: { "keyHandle": "KaWsrbTjPH1OB_MqvJCOw7SdeCQwbQUgx23TJ9UxYROIxSxXKyntggyccbO0sio8UYp6Ar4YrnUo6A40orkW6g", "version": "U2F_V2", "challenge": "aH2Hd1Tt3nYTzphPC-8oq1ny1F28H-bwLKwPyjbiYqc", "appId": "pam:\/\/grissom" }
JSON challenge URL-B64: aH2Hd1Tt3nYTzphPC-8oq1ny1F28H-bwLKwPyjbiYqc
client data: { "challenge": "aH2Hd1Tt3nYTzphPC-8oq1ny1F28H-bwLKwPyjbiYqc", "origin": "pam:\/\/grissom", "typ": "navigator.id.getAssertion" }
JSON: { "keyHandle": "KaWsrbTjPH1OB_MqvJCOw7SdeCQwbQUgx23TJ9UxYROIxSxXKyntggyccbO0sio8UYp6Ar4YrnUo6A40orkW6g", "version": "U2F_V2", "challenge": "aH2Hd1Tt3nYTzphPC-8oq1ny1F28H-bwLKwPyjbiYqc", "appId": "pam:\/\/grissom" }
JSON app_id pam://grissom
JSON: { "keyHandle": "KaWsrbTjPH1OB_MqvJCOw7SdeCQwbQUgx23TJ9UxYROIxSxXKyntggyccbO0sio8UYp6Ar4YrnUo6A40orkW6g", "version": "U2F_V2", "challenge": "aH2Hd1Tt3nYTzphPC-8oq1ny1F28H-bwLKwPyjbiYqc", "appId": "pam:\/\/grissom" }
JSON keyHandle URL-B64: KaWsrbTjPH1OB_MqvJCOw7SdeCQwbQUgx23TJ9UxYROIxSxXKyntggyccbO0sio8UYp6Ar4YrnUo6A40orkW6g
USB send: 0000c9000583008a00020700000081ea6edd8512262118b88616d35d5f8c6c0ff0b21943a9893bc0eb6b310faa89156240e70f034a89f360dac1a2c46c89b0a2
USB write returned 65
USB send: 0000c900050092aaf8eaac9d3076477ac75288a84029a5acadb4e33c7d4e07f32abc908ec3b49d7824306d0520c76dd327d531611388c52c572b29ed820c9c71
USB write returned 65
USB send: 0000c9000501b4b22a3c518a7a02be18ae7528e80e34a2b916ea0000000000000000000000000000000000000000000000000000000000000000000000000000
USB write returned 65
now trying with timeout 2
now trying with timeout 4
now trying with timeout 8
now trying with timeout 16
USB read rc read 64
USB recv: 00c900058300026a8000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000
USB data (len 2): 6a80
debug(pam_u2f): util.c:363 (do_authentication): Device for this keyhandle is not present.
USB send: 00ffffffff8600080807060504030201000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000
USB write returned 65
now trying with timeout 2
now trying with timeout 4
now trying with timeout 8
now trying with timeout 16
now trying with timeout 32
now trying with timeout 64
now trying with timeout 128
now trying with timeout 256
now trying with timeout 512
now trying with timeout 1024
now trying with timeout 2048
now trying with timeout 4096
USB read rc read 64
USB send: 0000c900058100010000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000
USB write returned 65
now trying with timeout 2
USB read rc read 64
USB recv: 00c90005810001000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000
debug(pam_u2f): util.c:379 (do_authentication): Devices max_index has changed: 3 (was 2). Starting over
debug(pam_u2f): util.c:306 (do_authentication): Attempting authentication with device number 1
debug(pam_u2f): util.c:330 (do_authentication): Challenge: { "keyHandle": "KaWsrbTjPH1OB_MqvJCOw7SdeCQwbQUgx23TJ9UxYROIxSxXKyntggyccbO0sio8UYp6Ar4YrnUo6A40orkW6g", "version": "U2F_V2", "challenge": "aH2Hd1Tt3nYTzphPC-8oq1ny1F28H-bwLKwPyjbiYqc", "appId": "pam:\/\/grissom" }
JSON: { "keyHandle": "KaWsrbTjPH1OB_MqvJCOw7SdeCQwbQUgx23TJ9UxYROIxSxXKyntggyccbO0sio8UYp6Ar4YrnUo6A40orkW6g", "version": "U2F_V2", "challenge": "aH2Hd1Tt3nYTzphPC-8oq1ny1F28H-bwLKwPyjbiYqc", "appId": "pam:\/\/grissom" }
JSON challenge URL-B64: aH2Hd1Tt3nYTzphPC-8oq1ny1F28H-bwLKwPyjbiYqc
client data: { "challenge": "aH2Hd1Tt3nYTzphPC-8oq1ny1F28H-bwLKwPyjbiYqc", "origin": "pam:\/\/grissom", "typ": "navigator.id.getAssertion" }
JSON: { "keyHandle": "KaWsrbTjPH1OB_MqvJCOw7SdeCQwbQUgx23TJ9UxYROIxSxXKyntggyccbO0sio8UYp6Ar4YrnUo6A40orkW6g", "version": "U2F_V2", "challenge": "aH2Hd1Tt3nYTzphPC-8oq1ny1F28H-bwLKwPyjbiYqc", "appId": "pam:\/\/grissom" }
JSON app_id pam://grissom
JSON: { "keyHandle": "KaWsrbTjPH1OB_MqvJCOw7SdeCQwbQUgx23TJ9UxYROIxSxXKyntggyccbO0sio8UYp6Ar4YrnUo6A40orkW6g", "version": "U2F_V2", "challenge": "aH2Hd1Tt3nYTzphPC-8oq1ny1F28H-bwLKwPyjbiYqc", "appId": "pam:\/\/grissom" }
JSON keyHandle URL-B64: KaWsrbTjPH1OB_MqvJCOw7SdeCQwbQUgx23TJ9UxYROIxSxXKyntggyccbO0sio8UYp6Ar4YrnUo6A40orkW6g
USB send: 0000c9000583008a00020700000081ea6edd8512262118b88616d35d5f8c6c0ff0b21943a9893bc0eb6b310faa89156240e70f034a89f360dac1a2c46c89b0a2
USB write returned 65
USB send: 0000c900050092aaf8eaac9d3076477ac75288a84029a5acadb4e33c7d4e07f32abc908ec3b49d7824306d0520c76dd327d531611388c52c572b29ed820c9c71
USB write returned 65
USB send: 0000c9000501b4b22a3c518a7a02be18ae7528e80e34a2b916ea0000000000000000000000000000000000000000000000000000000000000000000000000000
USB write returned 65
now trying with timeout 2
now trying with timeout 4
now trying with timeout 8
now trying with timeout 16
USB read rc read 64
USB recv: 00c900058300026a8000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000
USB data (len 2): 6a80
debug(pam_u2f): util.c:363 (do_authentication): Device for this keyhandle is not present.
USB send: 00ffffffff8600080807060504030201000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000
USB write returned 65
now trying with timeout 2
now trying with timeout 4
now trying with timeout 8
now trying with timeout 16
now trying with timeout 32
now trying with timeout 64
now trying with timeout 128
now trying with timeout 256
now trying with timeout 512
now trying with timeout 1024
now trying with timeout 2048
now trying with timeout 4096
USB read rc read 64
USB send: 0000c900058100010000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000
USB write returned 65
now trying with timeout 2
USB read rc read 64
USB recv: 00c90005810001000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000
debug(pam_u2f): util.c:379 (do_authentication): Devices max_index has changed: 4 (was 3). Starting over
debug(pam_u2f): util.c:306 (do_authentication): Attempting authentication with device number 1
debug(pam_u2f): util.c:330 (do_authentication): Challenge: { "keyHandle": "KaWsrbTjPH1OB_MqvJCOw7SdeCQwbQUgx23TJ9UxYROIxSxXKyntggyccbO0sio8UYp6Ar4YrnUo6A40orkW6g", "version": "U2F_V2", "challenge": "aH2Hd1Tt3nYTzphPC-8oq1ny1F28H-bwLKwPyjbiYqc", "appId": "pam:\/\/grissom" }
JSON: { "keyHandle": "KaWsrbTjPH1OB_MqvJCOw7SdeCQwbQUgx23TJ9UxYROIxSxXKyntggyccbO0sio8UYp6Ar4YrnUo6A40orkW6g", "version": "U2F_V2", "challenge": "aH2Hd1Tt3nYTzphPC-8oq1ny1F28H-bwLKwPyjbiYqc", "appId": "pam:\/\/grissom" }
JSON challenge URL-B64: aH2Hd1Tt3nYTzphPC-8oq1ny1F28H-bwLKwPyjbiYqc
client data: { "challenge": "aH2Hd1Tt3nYTzphPC-8oq1ny1F28H-bwLKwPyjbiYqc", "origin": "pam:\/\/grissom", "typ": "navigator.id.getAssertion" }
JSON: { "keyHandle": "KaWsrbTjPH1OB_MqvJCOw7SdeCQwbQUgx23TJ9UxYROIxSxXKyntggyccbO0sio8UYp6Ar4YrnUo6A40orkW6g", "version": "U2F_V2", "challenge": "aH2Hd1Tt3nYTzphPC-8oq1ny1F28H-bwLKwPyjbiYqc", "appId": "pam:\/\/grissom" }
JSON app_id pam://grissom
JSON: { "keyHandle": "KaWsrbTjPH1OB_MqvJCOw7SdeCQwbQUgx23TJ9UxYROIxSxXKyntggyccbO0sio8UYp6Ar4YrnUo6A40orkW6g", "version": "U2F_V2", "challenge": "aH2Hd1Tt3nYTzphPC-8oq1ny1F28H-bwLKwPyjbiYqc", "appId": "pam:\/\/grissom" }
JSON keyHandle URL-B64: KaWsrbTjPH1OB_MqvJCOw7SdeCQwbQUgx23TJ9UxYROIxSxXKyntggyccbO0sio8UYp6Ar4YrnUo6A40orkW6g
USB send: 0000c9000583008a00020700000081ea6edd8512262118b88616d35d5f8c6c0ff0b21943a9893bc0eb6b310faa89156240e70f034a89f360dac1a2c46c89b0a2
USB write returned 65
USB send: 0000c900050092aaf8eaac9d3076477ac75288a84029a5acadb4e33c7d4e07f32abc908ec3b49d7824306d0520c76dd327d531611388c52c572b29ed820c9c71
USB write returned 65
USB send: 0000c9000501b4b22a3c518a7a02be18ae7528e80e34a2b916ea0000000000000000000000000000000000000000000000000000000000000000000000000000
USB write returned 65
now trying with timeout 2
now trying with timeout 4
now trying with timeout 8
now trying with timeout 16
USB read rc read 64
USB recv: 00c900058300026a8000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000
USB data (len 2): 6a80
debug(pam_u2f): util.c:363 (do_authentication): Device for this keyhandle is not present.
USB send: 00ffffffff8600080807060504030201000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000
USB write returned 65
now trying with timeout 2
now trying with timeout 4
now trying with timeout 8
now trying with timeout 16
now trying with timeout 32
now trying with timeout 64
now trying with timeout 128
now trying with timeout 256
now trying with timeout 512
now trying with timeout 1024
now trying with timeout 2048
now trying with timeout 4096
USB read rc read 64
USB send: 0000c900058100010000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000
USB write returned 65
now trying with timeout 2
USB read rc read 64
USB recv: 00c90005810001000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000
debug(pam_u2f): util.c:379 (do_authentication): Devices max_index has changed: 5 (was 4). Starting over
debug(pam_u2f): util.c:306 (do_authentication): Attempting authentication with device number 1
debug(pam_u2f): util.c:330 (do_authentication): Challenge: { "keyHandle": "KaWsrbTjPH1OB_MqvJCOw7SdeCQwbQUgx23TJ9UxYROIxSxXKyntggyccbO0sio8UYp6Ar4YrnUo6A40orkW6g", "version": "U2F_V2", "challenge": "aH2Hd1Tt3nYTzphPC-8oq1ny1F28H-bwLKwPyjbiYqc", "appId": "pam:\/\/grissom" }
JSON: { "keyHandle": "KaWsrbTjPH1OB_MqvJCOw7SdeCQwbQUgx23TJ9UxYROIxSxXKyntggyccbO0sio8UYp6Ar4YrnUo6A40orkW6g", "version": "U2F_V2", "challenge": "aH2Hd1Tt3nYTzphPC-8oq1ny1F28H-bwLKwPyjbiYqc", "appId": "pam:\/\/grissom" }
JSON challenge URL-B64: aH2Hd1Tt3nYTzphPC-8oq1ny1F28H-bwLKwPyjbiYqc
client data: { "challenge": "aH2Hd1Tt3nYTzphPC-8oq1ny1F28H-bwLKwPyjbiYqc", "origin": "pam:\/\/grissom", "typ": "navigator.id.getAssertion" }
JSON: { "keyHandle": "KaWsrbTjPH1OB_MqvJCOw7SdeCQwbQUgx23TJ9UxYROIxSxXKyntggyccbO0sio8UYp6Ar4YrnUo6A40orkW6g", "version": "U2F_V2", "challenge": "aH2Hd1Tt3nYTzphPC-8oq1ny1F28H-bwLKwPyjbiYqc", "appId": "pam:\/\/grissom" }
JSON app_id pam://grissom
JSON: { "keyHandle": "KaWsrbTjPH1OB_MqvJCOw7SdeCQwbQUgx23TJ9UxYROIxSxXKyntggyccbO0sio8UYp6Ar4YrnUo6A40orkW6g", "version": "U2F_V2", "challenge": "aH2Hd1Tt3nYTzphPC-8oq1ny1F28H-bwLKwPyjbiYqc", "appId": "pam:\/\/grissom" }
JSON keyHandle URL-B64: KaWsrbTjPH1OB_MqvJCOw7SdeCQwbQUgx23TJ9UxYROIxSxXKyntggyccbO0sio8UYp6Ar4YrnUo6A40orkW6g
USB send: 0000c9000583008a00020700000081ea6edd8512262118b88616d35d5f8c6c0ff0b21943a9893bc0eb6b310faa89156240e70f034a89f360dac1a2c46c89b0a2
USB write returned 65
USB send: 0000c900050092aaf8eaac9d3076477ac75288a84029a5acadb4e33c7d4e07f32abc908ec3b49d7824306d0520c76dd327d531611388c52c572b29ed820c9c71
USB write returned 65
USB send: 0000c9000501b4b22a3c518a7a02be18ae7528e80e34a2b916ea0000000000000000000000000000000000000000000000000000000000000000000000000000
USB write returned 65
now trying with timeout 2
now trying with timeout 4
now trying with timeout 8
now trying with timeout 16
USB read rc read 64
USB recv: 00c900058300026a8000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000
USB data (len 2): 6a80
debug(pam_u2f): util.c:363 (do_authentication): Device for this keyhandle is not present.
USB send: 00ffffffff8600080807060504030201000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000
USB write returned 65
now trying with timeout 2
now trying with timeout 4
now trying with timeout 8
now trying with timeout 16
now trying with timeout 32
now trying with timeout 64
now trying with timeout 128
now trying with timeout 256
now trying with timeout 512
now trying with timeout 1024
now trying with timeout 2048
now trying with timeout 4096
USB read rc read 64
USB send: 0000c900058100010000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000
USB write returned 65
now trying with timeout 2
USB read rc read 64
USB recv: 00c90005810001000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000
debug(pam_u2f): util.c:379 (do_authentication): Devices max_index has changed: 6 (was 5). Starting over
debug(pam_u2f): util.c:306 (do_authentication): Attempting authentication with device number 1
debug(pam_u2f): util.c:330 (do_authentication): Challenge: { "keyHandle": "KaWsrbTjPH1OB_MqvJCOw7SdeCQwbQUgx23TJ9UxYROIxSxXKyntggyccbO0sio8UYp6Ar4YrnUo6A40orkW6g", "version": "U2F_V2", "challenge": "aH2Hd1Tt3nYTzphPC-8oq1ny1F28H-bwLKwPyjbiYqc", "appId": "pam:\/\/grissom" }
JSON: { "keyHandle": "KaWsrbTjPH1OB_MqvJCOw7SdeCQwbQUgx23TJ9UxYROIxSxXKyntggyccbO0sio8UYp6Ar4YrnUo6A40orkW6g", "version": "U2F_V2", "challenge": "aH2Hd1Tt3nYTzphPC-8oq1ny1F28H-bwLKwPyjbiYqc", "appId": "pam:\/\/grissom" }
JSON challenge URL-B64: aH2Hd1Tt3nYTzphPC-8oq1ny1F28H-bwLKwPyjbiYqc
client data: { "challenge": "aH2Hd1Tt3nYTzphPC-8oq1ny1F28H-bwLKwPyjbiYqc", "origin": "pam:\/\/grissom", "typ": "navigator.id.getAssertion" }
JSON: { "keyHandle": "KaWsrbTjPH1OB_MqvJCOw7SdeCQwbQUgx23TJ9UxYROIxSxXKyntggyccbO0sio8UYp6Ar4YrnUo6A40orkW6g", "version": "U2F_V2", "challenge": "aH2Hd1Tt3nYTzphPC-8oq1ny1F28H-bwLKwPyjbiYqc", "appId": "pam:\/\/grissom" }
JSON app_id pam://grissom
JSON: { "keyHandle": "KaWsrbTjPH1OB_MqvJCOw7SdeCQwbQUgx23TJ9UxYROIxSxXKyntggyccbO0sio8UYp6Ar4YrnUo6A40orkW6g", "version": "U2F_V2", "challenge": "aH2Hd1Tt3nYTzphPC-8oq1ny1F28H-bwLKwPyjbiYqc", "appId": "pam:\/\/grissom" }
JSON keyHandle URL-B64: KaWsrbTjPH1OB_MqvJCOw7SdeCQwbQUgx23TJ9UxYROIxSxXKyntggyccbO0sio8UYp6Ar4YrnUo6A40orkW6g
USB send: 0000c9000583008a00020700000081ea6edd8512262118b88616d35d5f8c6c0ff0b21943a9893bc0eb6b310faa89156240e70f034a89f360dac1a2c46c89b0a2
USB write returned 65
USB send: 0000c900050092aaf8eaac9d3076477ac75288a84029a5acadb4e33c7d4e07f32abc908ec3b49d7824306d0520c76dd327d531611388c52c572b29ed820c9c71
USB write returned 65
USB send: 0000c9000501b4b22a3c518a7a02be18ae7528e80e34a2b916ea0000000000000000000000000000000000000000000000000000000000000000000000000000
USB write returned 65
now trying with timeout 2
now trying with timeout 4
now trying with timeout 8
now trying with timeout 16
USB read rc read 64
USB recv: 00c900058300026a8000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000
USB data (len 2): 6a80
debug(pam_u2f): util.c:363 (do_authentication): Device for this keyhandle is not present.
USB send: 00ffffffff8600080807060504030201000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000
USB write returned 65
now trying with timeout 2
now trying with timeout 4
now trying with timeout 8
now trying with timeout 16
now trying with timeout 32
now trying with timeout 64
now trying with timeout 128
now trying with timeout 256
now trying with timeout 512
now trying with timeout 1024
now trying with timeout 2048
now trying with timeout 4096
`

klali commented:
That's annoying.. For me it seems to work as expected with the index not increasing after a failed device. I've added another commit with a bit more debug information on that branch, can I convince you to test again with a fresh build of it?

cjoster commented:
Sure, no problem. This token is really poorly behaved, so this might be all for naught. For example, it (almost) never rejects a keyhandle and has no notion of a check-only authentication, so this might not be u2f-host's problem, but, it does break authentication just by plugging one in (configured or not). [See here.](https://github.com/trustcrypto/OnlyKey-Firmware/issues/75)

```
This tty (pts/4) is not a virtual console.


The pts/4 is now locked by lordvadr.
Password: 
debug(pam_u2f): pam-u2f.c:94 (parse_cfg): called.
debug(pam_u2f): pam-u2f.c:95 (parse_cfg): flags 0 argc 3
debug(pam_u2f): pam-u2f.c:97 (parse_cfg): argv[0]=cue
debug(pam_u2f): pam-u2f.c:97 (parse_cfg): argv[1]=authpending_file=
debug(pam_u2f): pam-u2f.c:97 (parse_cfg): argv[2]=debug
debug(pam_u2f): pam-u2f.c:99 (parse_cfg): max_devices=0
debug(pam_u2f): pam-u2f.c:100 (parse_cfg): debug=1
debug(pam_u2f): pam-u2f.c:101 (parse_cfg): interactive=0
debug(pam_u2f): pam-u2f.c:102 (parse_cfg): cue=1
debug(pam_u2f): pam-u2f.c:103 (parse_cfg): nodetect=0
debug(pam_u2f): pam-u2f.c:104 (parse_cfg): manual=0
debug(pam_u2f): pam-u2f.c:105 (parse_cfg): nouserok=0
debug(pam_u2f): pam-u2f.c:106 (parse_cfg): openasuser=0
debug(pam_u2f): pam-u2f.c:107 (parse_cfg): alwaysok=0
debug(pam_u2f): pam-u2f.c:108 (parse_cfg): authfile=(null)
debug(pam_u2f): pam-u2f.c:109 (parse_cfg): authpending_file=
debug(pam_u2f): pam-u2f.c:110 (parse_cfg): origin=(null)
debug(pam_u2f): pam-u2f.c:111 (parse_cfg): appid=(null)
debug(pam_u2f): pam-u2f.c:112 (parse_cfg): prompt=(null)
debug(pam_u2f): pam-u2f.c:157 (pam_sm_authenticate): Origin not specified, using "pam://grissom"
debug(pam_u2f): pam-u2f.c:169 (pam_sm_authenticate): Appid not specified, using the same value of origin (pam://grissom)
debug(pam_u2f): pam-u2f.c:180 (pam_sm_authenticate): Maximum devices number not set. Using default (24)
debug(pam_u2f): pam-u2f.c:198 (pam_sm_authenticate): Requesting authentication for user lordvadr
debug(pam_u2f): pam-u2f.c:209 (pam_sm_authenticate): Found user lordvadr
debug(pam_u2f): pam-u2f.c:210 (pam_sm_authenticate): Home directory for lordvadr is /home/lordvadr
debug(pam_u2f): pam-u2f.c:217 (pam_sm_authenticate): Variable XDG_CONFIG_HOME is not set. Using default value ($HOME/.config/)
debug(pam_u2f): pam-u2f.c:245 (pam_sm_authenticate): Using default authentication file /home/lordvadr/.config/Yubico/u2f_keys
debug(pam_u2f): util.c:102 (get_devices_from_authfile): Authorization line: lordvadr:KaWsrbTjPH1OB_MqvJCOw7SdeCQwbQUgx23TJ9UxYROIxSxXKyntggyccbO0sio8UYp6Ar4YrnUo6A40orkW6g,04da9404e660e720f5cca1645a54847d5ed844e1ea2366af025dcb98cd22e9fbce42104003714c3bc3f53ad42ef9993b23523bc26b564e139aa93f7769a5c26df0:uON-6lDuYfmJOeYDqI8q4ZR-eEFsRAStxKorGd6BaUV6MxIBRDB8JdGKP0j5ysNavEfVMPCPb0MwSNJ5WUB6Lw,04b1393e667f66da06041ad4081cbc0b681dee1d5e5b70dcbe95e80115d71a66423fb34ea701bf04c738a358f596a4a1ce49cd15c5e29aba38559a47daab48512f:PqSNe6j2l_ConOy4H0YmPLdVQ2zOOr6Xw5bkFp17Xh5CXOZeTv614p2FahQOYXLk2GRdWwF1L1nbxbgbB55A3g,04a0f65042cd6c3bd6e25ae7d0383cb3fdd722c3e576df892ae5706e0b2a2a7532cae55c4194db6aef489ebe44d2f6e20c8cdaca63bb74340d0af33557d1bdf447
debug(pam_u2f): util.c:107 (get_devices_from_authfile): Matched user: lordvadr
debug(pam_u2f): util.c:134 (get_devices_from_authfile): KeyHandle for device number 1: KaWsrbTjPH1OB_MqvJCOw7SdeCQwbQUgx23TJ9UxYROIxSxXKyntggyccbO0sio8UYp6Ar4YrnUo6A40orkW6g
debug(pam_u2f): util.c:153 (get_devices_from_authfile): publicKey for device number 1: 04da9404e660e720f5cca1645a54847d5ed844e1ea2366af025dcb98cd22e9fbce42104003714c3bc3f53ad42ef9993b23523bc26b564e139aa93f7769a5c26df0
debug(pam_u2f): util.c:164 (get_devices_from_authfile): Length of key number 1 is 65
debug(pam_u2f): util.c:134 (get_devices_from_authfile): KeyHandle for device number 2: uON-6lDuYfmJOeYDqI8q4ZR-eEFsRAStxKorGd6BaUV6MxIBRDB8JdGKP0j5ysNavEfVMPCPb0MwSNJ5WUB6Lw
debug(pam_u2f): util.c:153 (get_devices_from_authfile): publicKey for device number 2: 04b1393e667f66da06041ad4081cbc0b681dee1d5e5b70dcbe95e80115d71a66423fb34ea701bf04c738a358f596a4a1ce49cd15c5e29aba38559a47daab48512f
debug(pam_u2f): util.c:164 (get_devices_from_authfile): Length of key number 2 is 65
debug(pam_u2f): util.c:134 (get_devices_from_authfile): KeyHandle for device number 3: PqSNe6j2l_ConOy4H0YmPLdVQ2zOOr6Xw5bkFp17Xh5CXOZeTv614p2FahQOYXLk2GRdWwF1L1nbxbgbB55A3g
debug(pam_u2f): util.c:153 (get_devices_from_authfile): publicKey for device number 3: 04a0f65042cd6c3bd6e25ae7d0383cb3fdd722c3e576df892ae5706e0b2a2a7532cae55c4194db6aef489ebe44d2f6e20c8cdaca63bb74340d0af33557d1bdf447
debug(pam_u2f): util.c:164 (get_devices_from_authfile): Length of key number 3 is 65
debug(pam_u2f): util.c:102 (get_devices_from_authfile): Authorization line: :HQJH5olY_fA4qkz-NZot0pr_ZmqFEzzv4rj2J6OiChRRivHa9bwb1xKNaJe0uS8l4B0bKynkoF0MS-rz6dmaNQ,04a96b010043fafd46b67a95eae08e607ce87824633903ca0dd374a57db11ba2927bbad311f5c481739ae7deae06ff7f077c45c32a9d983e1c5c3c9948642b8874
debug(pam_u2f): util.c:191 (get_devices_from_authfile): Found 3 device(s) for user lordvadr
debug(pam_u2f): pam-u2f.c:310 (pam_sm_authenticate): authpending_file is set to an empty value, touch request notifications will be disabled
USB send: 00ffffffff8600080807060504030201000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000
USB write returned 65
now trying with timeout 2
now trying with timeout 4
now trying with timeout 8
now trying with timeout 16
now trying with timeout 32
now trying with timeout 64
now trying with timeout 128
now trying with timeout 256
now trying with timeout 512
USB read rc read 64
USB recv: 494e495449414c495a45440000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000
USB send: 00ffffffff8600080807060504030201000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000
USB write returned 65
now trying with timeout 2
USB read rc read 64
USB recv: ffffffff860011080706050403020100dc0003020403070100000000000000000000000000000000000000000000000000000000000000000000000000000000
device /dev/hidraw1 discovered as 'Yubikey 4 OTP+U2F+CCID'
  version (Interface, Major, Minor, Build): 2, 4, 3, 0  capFlags: 1
debug(pam_u2f): util.c:272 (do_authentication): Device max index is 1
debug(pam_u2f): util.c:306 (do_authentication): Attempting authentication with device number 1
debug(pam_u2f): util.c:330 (do_authentication): Challenge: { "keyHandle": "KaWsrbTjPH1OB_MqvJCOw7SdeCQwbQUgx23TJ9UxYROIxSxXKyntggyccbO0sio8UYp6Ar4YrnUo6A40orkW6g", "version": "U2F_V2", "challenge": "qQjauzKuT33JD8n-eJv4ISBdNPWNGFHHkIAqWzH6LpU", "appId": "pam:\/\/grissom" }
JSON: { "keyHandle": "KaWsrbTjPH1OB_MqvJCOw7SdeCQwbQUgx23TJ9UxYROIxSxXKyntggyccbO0sio8UYp6Ar4YrnUo6A40orkW6g", "version": "U2F_V2", "challenge": "qQjauzKuT33JD8n-eJv4ISBdNPWNGFHHkIAqWzH6LpU", "appId": "pam:\/\/grissom" }
JSON challenge URL-B64: qQjauzKuT33JD8n-eJv4ISBdNPWNGFHHkIAqWzH6LpU
client data: { "challenge": "qQjauzKuT33JD8n-eJv4ISBdNPWNGFHHkIAqWzH6LpU", "origin": "pam:\/\/grissom", "typ": "navigator.id.getAssertion" }
JSON: { "keyHandle": "KaWsrbTjPH1OB_MqvJCOw7SdeCQwbQUgx23TJ9UxYROIxSxXKyntggyccbO0sio8UYp6Ar4YrnUo6A40orkW6g", "version": "U2F_V2", "challenge": "qQjauzKuT33JD8n-eJv4ISBdNPWNGFHHkIAqWzH6LpU", "appId": "pam:\/\/grissom" }
JSON app_id pam://grissom
JSON: { "keyHandle": "KaWsrbTjPH1OB_MqvJCOw7SdeCQwbQUgx23TJ9UxYROIxSxXKyntggyccbO0sio8UYp6Ar4YrnUo6A40orkW6g", "version": "U2F_V2", "challenge": "qQjauzKuT33JD8n-eJv4ISBdNPWNGFHHkIAqWzH6LpU", "appId": "pam:\/\/grissom" }
JSON keyHandle URL-B64: KaWsrbTjPH1OB_MqvJCOw7SdeCQwbQUgx23TJ9UxYROIxSxXKyntggyccbO0sio8UYp6Ar4YrnUo6A40orkW6g
USB send: 0000dc000383008a00020700000081d50ee2f91e1f2a5b11143533ca63ee15c9a6623b4b6ad09df0380ffb1f17ff576240e70f034a89f360dac1a2c46c89b0a2
USB write returned 65
USB send: 0000dc00030092aaf8eaac9d3076477ac75288a84029a5acadb4e33c7d4e07f32abc908ec3b49d7824306d0520c76dd327d531611388c52c572b29ed820c9c71
USB write returned 65
USB send: 0000dc000301b4b22a3c518a7a02be18ae7528e80e34a2b916ea0000000000000000000000000000000000000000000000000000000000000000000000000000
USB write returned 65
now trying with timeout 2
now trying with timeout 4
now trying with timeout 8
now trying with timeout 16
USB read rc read 64
USB recv: 00dc00038300026a8000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000
USB data (len 2): 6a80
debug(pam_u2f): util.c:363 (do_authentication): Device for this keyhandle is not present.
USB send: 00ffffffff8600080807060504030201000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000
USB write returned 65
now trying with timeout 2
now trying with timeout 4
now trying with timeout 8
now trying with timeout 16
now trying with timeout 32
now trying with timeout 64
now trying with timeout 128
now trying with timeout 256
now trying with timeout 512
USB read rc read 64
USB recv: 494e495449414c495a45440000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000
USB send: 0000dc00038100010000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000
USB write returned 65
now trying with timeout 2
USB read rc read 64
USB recv: 00dc0003810001000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000
debug(pam_u2f): util.c:379 (do_authentication): Devices max_index has changed: 2 (was 1). Starting over
debug(pam_u2f): util.c:306 (do_authentication): Attempting authentication with device number 1
debug(pam_u2f): util.c:330 (do_authentication): Challenge: { "keyHandle": "KaWsrbTjPH1OB_MqvJCOw7SdeCQwbQUgx23TJ9UxYROIxSxXKyntggyccbO0sio8UYp6Ar4YrnUo6A40orkW6g", "version": "U2F_V2", "challenge": "qQjauzKuT33JD8n-eJv4ISBdNPWNGFHHkIAqWzH6LpU", "appId": "pam:\/\/grissom" }
JSON: { "keyHandle": "KaWsrbTjPH1OB_MqvJCOw7SdeCQwbQUgx23TJ9UxYROIxSxXKyntggyccbO0sio8UYp6Ar4YrnUo6A40orkW6g", "version": "U2F_V2", "challenge": "qQjauzKuT33JD8n-eJv4ISBdNPWNGFHHkIAqWzH6LpU", "appId": "pam:\/\/grissom" }
JSON challenge URL-B64: qQjauzKuT33JD8n-eJv4ISBdNPWNGFHHkIAqWzH6LpU
client data: { "challenge": "qQjauzKuT33JD8n-eJv4ISBdNPWNGFHHkIAqWzH6LpU", "origin": "pam:\/\/grissom", "typ": "navigator.id.getAssertion" }
JSON: { "keyHandle": "KaWsrbTjPH1OB_MqvJCOw7SdeCQwbQUgx23TJ9UxYROIxSxXKyntggyccbO0sio8UYp6Ar4YrnUo6A40orkW6g", "version": "U2F_V2", "challenge": "qQjauzKuT33JD8n-eJv4ISBdNPWNGFHHkIAqWzH6LpU", "appId": "pam:\/\/grissom" }
JSON app_id pam://grissom
JSON: { "keyHandle": "KaWsrbTjPH1OB_MqvJCOw7SdeCQwbQUgx23TJ9UxYROIxSxXKyntggyccbO0sio8UYp6Ar4YrnUo6A40orkW6g", "version": "U2F_V2", "challenge": "qQjauzKuT33JD8n-eJv4ISBdNPWNGFHHkIAqWzH6LpU", "appId": "pam:\/\/grissom" }
JSON keyHandle URL-B64: KaWsrbTjPH1OB_MqvJCOw7SdeCQwbQUgx23TJ9UxYROIxSxXKyntggyccbO0sio8UYp6Ar4YrnUo6A40orkW6g
USB send: 0000dc000383008a00020700000081d50ee2f91e1f2a5b11143533ca63ee15c9a6623b4b6ad09df0380ffb1f17ff576240e70f034a89f360dac1a2c46c89b0a2
USB write returned 65
USB send: 0000dc00030092aaf8eaac9d3076477ac75288a84029a5acadb4e33c7d4e07f32abc908ec3b49d7824306d0520c76dd327d531611388c52c572b29ed820c9c71
USB write returned 65
USB send: 0000dc000301b4b22a3c518a7a02be18ae7528e80e34a2b916ea0000000000000000000000000000000000000000000000000000000000000000000000000000
USB write returned 65
now trying with timeout 2
now trying with timeout 4
now trying with timeout 8
now trying with timeout 16
USB read rc read 64
USB recv: 00dc00038300026a8000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000
USB data (len 2): 6a80
debug(pam_u2f): util.c:363 (do_authentication): Device for this keyhandle is not present.
USB send: 00ffffffff8600080807060504030201000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000
USB write returned 65
now trying with timeout 2
now trying with timeout 4
now trying with timeout 8
now trying with timeout 16
now trying with timeout 32
now trying with timeout 64
now trying with timeout 128
now trying with timeout 256
now trying with timeout 512
USB read rc read 64
USB recv: 494e495449414c495a45440000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000
USB send: 0000dc00038100010000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000
USB write returned 65
now trying with timeout 2
USB read rc read 64
USB recv: 00dc0003810001000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000
debug(pam_u2f): util.c:379 (do_authentication): Devices max_index has changed: 3 (was 2). Starting over
```

nevun commented:
We are archiving this project and closing PRs. Please open a new one in the replacement library's repo: https://github.com/Yubico/libfido if it is still applicable.
